contracts/zero-ex/contracts/src/features/ERC721OrdersFeature.sol

```diff
@@ -0,0 +1,1112 @@
+        uint256 ethAvailable,
+        bytes memory takerCallbackData
+    )
+        public
+        payable
+        returns (uint256 ethSpent)
+    {
+        // Check that the order can be filled.
+        _validateSellOrder(order, signature);
+
+        // Mark the order as filled.
+        _setOrderStatusBit(order.maker, order.nonce);
+
+        // Transfer the ERC721 asset to the buyer (`msg.sender`).
+        _transferERC721AssetFrom(
+            order.erc721Token,
+            order.maker,
+            msg.sender,
+            order.erc721TokenId
+        );
+
+        if (takerCallbackData.length > 0) {
+            require(
+                msg.sender != address(this), 
+                "ERC721OrdersFeature::_buyERC721/CANNOT_CALLBACK_SELF"
+            );
+            uint256 ethBalanceBeforeCallback = address(this).balance;
+            // Invoke the callback
+            bytes4 callbackResult = IERC721OrderCallback(msg.sender)
+                .zeroExERC721OrderCallback(takerCallbackData);
+            ethAvailable = ethAvailable.safeAdd(
+                address(this).balance.safeSub(ethBalanceBeforeCallback)
+            );
+            // Check for the magic success bytes
+            require(
+                callbackResult == TAKER_CALLBACK_MAGIC_BYTES,
+                "ERC721OrdersFeature::_sellERC721/CALLBACK_FAILED"
+            );            
+        }
+
+        if (address(order.erc20Token) == NATIVE_TOKEN_ADDRESS) {
+            // Check that we have enough ETH.
+            rrequire(
+                ethAvailable >= order.erc20TokenAmount,
+                LibERC721OrdersRichErrors.InsufficientEthError(
+                    ethAvailable,
+                    order.erc20TokenAmount
+                )
+            );
+            // Transfer ETH to the seller.
+            _transferEth(payable(order.maker), order.erc20TokenAmount);
+            // Fees are paid from the EP's current balance of ETH.
+            uint256 ethFees = _payFees(order, address(this), true);
+            // Sum the amount of ETH spent.
+            ethSpent = order.erc20TokenAmount.safeAdd(ethFees);
+        } else if (order.erc20Token == WETH) {
+            // If there is enough ETH available, fill the WETH order 
+            // (including fees) using that ETH. 
+            // Otherwise, transfer WETH from the taker.
+            if (ethAvailable >= order.erc20TokenAmount) {
+                // Wrap ETH.
+                WETH.deposit{value: order.erc20TokenAmount}();
+                // TODO: Probably safe to just use WETH.transfer for some
+                //       small gas savings
+                // Transfer WETH to the seller.
+                _transferERC20Tokens(
+                    WETH,
+                    order.maker,
+                    order.erc20TokenAmount
+                );
+                // Pay fees using ETH. 
+                uint256 ethFees = _payFees(order, address(this), true);
```

Comment:
`ethAvailable >= order.erc20TokenAmount` doesn't account for fees though? So it's possible this'll run out?

Reply:
idk should `_payFees()` intelligently wrap on the fly? maybe we write a smart `transferTokens()` function that wraps on the fly and handles all the edge cases? Is that possible?

Reply:
Or I guess we're OK with saying that if you're filling with ETH then you must supply enough to cover the fees as well. Also, I know we also discussed this before but it feels worse in person that fees can come in either ETH or WETH at the mercy of the taker. At the very least we should call out that we've accepted this edge case with a comment for the auditors.

Reply:
Yeah, we ran this by some potential integrators and they seem to be ok with the fee recipients potentially receiving a mix of ETH/WETH. But yeah we can specifically call this out for auditors